tests/hikari/impl/test_rest.py

```diff
@@ -1557,6 +1558,90 @@ def test__build_message_payload_when_both_single_and_plural_args_passed(
+        component = mock.Mock(type=type, build=mock.Mock(return_value=({}, ())))
+
+        payload, _ = rest_client._build_message_payload(component=component)
+
+        assert payload.get("flags") is None
+
+    @pytest.mark.parametrize(
+        "type",
+        (
+            components.ComponentType.SECTION,
+            components.ComponentType.TEXT_DISPLAY,
+            components.ComponentType.THUMBNAIL,
+            components.ComponentType.MEDIA_GALLERY,
+            components.ComponentType.FILE,
+            components.ComponentType.SEPARATOR,
+            components.ComponentType.CONTAINER,
+        ),
+    )
+    def test__build_message_payload_with_singular_components_v2(self, rest_client, type):
+        component = mock.Mock(type=type, build=mock.Mock(return_value=({}, ())))
+
+        payload, _ = rest_client._build_message_payload(component=component)
+
+        assert payload.get("flags") is message_models.MessageFlag.IS_COMPONENTS_V2
+
+    @pytest.mark.parametrize(
+        "type",
+        (
+            components.ComponentType.ACTION_ROW,
+            components.ComponentType.BUTTON,
+            components.ComponentType.TEXT_SELECT_MENU,
+            components.ComponentType.USER_SELECT_MENU,
+            components.ComponentType.ROLE_SELECT_MENU,
+            components.ComponentType.MENTIONABLE_SELECT_MENU,
+            components.ComponentType.CHANNEL_SELECT_MENU,
+        ),
+    )
+    def test__build_message_payload_with_multiple_components_v1(self, rest_client, type):
+        component = mock.Mock(type=type, build=mock.Mock(return_value=({}, ())))
+
+        payload, _ = rest_client._build_message_payload(component=component)
+
+        assert payload.get("flags") is None
+
+    @pytest.mark.parametrize(
+        "type",
+        (
+            components.ComponentType.SECTION,
+            components.ComponentType.TEXT_DISPLAY,
+            components.ComponentType.THUMBNAIL,
+            components.ComponentType.MEDIA_GALLERY,
+            components.ComponentType.FILE,
+            components.ComponentType.SEPARATOR,
+            components.ComponentType.CONTAINER,
+        ),
+    )
+    def test__build_message_payload_with_multiple_components_v2(self, rest_client, type):
+        component = mock.Mock(type=type, build=mock.Mock(return_value=({}, ())))
+
+        payload, _ = rest_client._build_message_payload(components=[component])
+
+        assert payload.get("flags") is message_models.MessageFlag.IS_COMPONENTS_V2
+
+    def test__build_message_payload_with_mixed_components(self, rest_client):
+        component_1 = mock.Mock(type=components.ComponentType.ACTION_ROW, build=mock.Mock(return_value=({}, ())))
+        component_2 = mock.Mock(type=components.ComponentType.CONTAINER, build=mock.Mock(return_value=({}, ())))
+
+        payload, _ = rest_client._build_message_payload(components=[component_1, component_2])
+
+        assert payload.get("flags") is message_models.MessageFlag.IS_COMPONENTS_V2
+
```

Comment:
Would be nice to also have an extra test to ensure that if the user passes flags, we do not override them, and instead just build on top of them